extensions/ql-vscode/src/queries-panel/queries-module.ts

```diff
@@ -6,12 +6,42 @@ import { DisposableObject } from "../common/disposable-object";
 import { QueriesPanel } from "./queries-panel";
 import { QueryDiscovery } from "./query-discovery";
 import { QueryPackDiscovery } from "./query-pack-discovery";
+import { QueriesPanelCommands } from "../common/commands";
 
 export class QueriesModule extends DisposableObject {
+  private queriesPanel: QueriesPanel | undefined;
+
   private constructor(readonly app: App) {
     super();
   }
 
+  public static initialize(
+    app: App,
+    cliServer: CodeQLCliServer,
+  ): QueriesModule {
+    const queriesModule = new QueriesModule(app);
+    app.subscriptions.push(queriesModule);
+
+    queriesModule.initialize(app, cliServer);
+    return queriesModule;
+  }
+
+  public getCommands(): QueriesPanelCommands {
+    if (!isCanary() || !showQueriesPanel()) {
+      // Currently, we only want to expose the new panel when we are in development and canary mode
+      // and the developer has enabled the "Show queries panel" flag.
+      return {} as any as QueriesPanelCommands;
```

Comment:
I don't particularly love that we have to do this cast. We shouldn't have to force the type system, so I think it shows that things perhaps aren't quite right at a high-level.

Earlier today we were discussing always registering the command and then only showing the panel if the feature flag is enabled. That would mean something like https://github.com/github/vscode-codeql/commit/42886c69aed49878d66a4e298e0a6829de00cee7, though perhaps I didn't make it clear enough what I meant. What do you think? If you don't think that's better then that's ok, but I'd be interested to hear the reasoning.

Or as another alternative would it be better providing dummy implementations that just log an error? That way if one of these commands does accidentally become callable then the user will get a nicer error message.

Reply:
Ah, yes. My thought was to initialize as little as possible. I agree that the cast is a code smell, but to push it all to the panel and initialize that didn't sit well with me either.

Reply:
We have some other optional commands as well, the `TestUICommands` commands. These are declared as `Partial<TestUICommands>` in the `commands.ts` file so that they aren't required to be set. I think that would work well here as well.